src/main/java/uk/gov/pay/directdebit/payments/dao/DirectDebitEventDao.java

```diff
@@ -1,26 +1,114 @@
+        return jdbi.withHandle(handle -> {
+            Query query = handle.createQuery(QUERY
+                    .replace(":searchFields", queryStringAndQueryMap.queryString)
+                    .replace(":limit", limit)
+                    .replace(":offset", offset));
+            queryStringAndQueryMap.queryMap.forEach(query::bind);
+            return query.map(new EventMapper()).list();
+        });
+    }
+
+    private QueryStringAndQueryMap generateQuery(DirectDebitEventSearchParams searchParams) {
+        List<String> searchStrings = Lists.newArrayList();
+        Map<String, Object> queryMap = Maps.newHashMap();
+        if (nonNull(searchParams.getMandateExternalId())) {
+            searchStrings.add("m.external_id = :mandate_id");
+            queryMap.put("mandate_id", searchParams.getMandateExternalId());
+        }
+        if (nonNull(searchParams.getTransactionExternalId())) {
+            searchStrings.add("t.external_id = :transaction_id");
+            queryMap.put("transaction_id", searchParams.getTransactionExternalId());
+        }
+        if (nonNull(searchParams.getBeforeDate())) {
+            searchStrings.add("e.event_date < :before_date");
+            queryMap.put("before_date", searchParams.getBeforeDate());
+        }
+        if (nonNull(searchParams.getAfterDate())) {
+            searchStrings.add("e.event_date > :after_date");
+            queryMap.put("after_date", searchParams.getAfterDate());
+        }
+        String queryString = searchStrings.isEmpty() ? "" : "WHERE " + searchStrings.stream().collect(Collectors.joining(" AND "));
+        return new QueryStringAndQueryMap(queryString, queryMap);
+    }
+
+    public int getTotalNumberOfEvents(DirectDebitEventSearchParams searchParams) {
+        QueryStringAndQueryMap queryStringAndQueryMap = generateQuery(searchParams);
+        return jdbi.withHandle(handle -> {
+            Query query = handle.createQuery(COUNT_QUERY.replace(":searchFields", queryStringAndQueryMap.queryString));
+            queryStringAndQueryMap.queryMap.forEach(query::bind);
+            return query
+                    .mapTo(Integer.class)
+                    .findOnly();
+        });
+    }
+    
+    public Long insert(DirectDebitEvent directDebitEvent) {
+        return (long) jdbi.withHandle(handle -> {
+            return handle.createUpdate("INSERT INTO events(mandate_id, external_id, transaction_id, event_type, " +
+                    "event, event_date) VALUES (:mandateId, :externalId, :transactionId, :eventType, :event, :eventDate)")
+                    .bindBean(directDebitEvent)
+                    .execute();
+        });
+    }
+
+    public Optional<DirectDebitEvent> findByMandateIdAndEvent(Long mandateId, DirectDebitEvent.Type eventType, DirectDebitEvent.SupportedEvent event) {
+        return jdbi.withHandle(handle -> handle.createQuery("SELECT e.id, e.mandate_id, e.transaction_id, e.event_type, e.event, e.event_date, e.version, e.external_id, " +
```

Comment:
Can you not use the `QUERY` constant here?

Reply:
not really. they are subtly different. The constant `QUERY`'s searchFields have a dynamic WHERE clause, whereas the query here does not.